### ISSUE TITLE
downloading junk messages raise socket error with "copy message list.remove(x): x not in list" 

### ISSUE THREAD

**Issue by naveen1336** [OfflineIMAP/offlineimap#198]
offlineimap v 6.5.7 rc4 latest
this error occur very frequently my configuration is normal 
![error](https://cloud.githubusercontent.com/assets/8940514/7631471/5367bd6e-fa60-11e4-9cfb-a6f1d52efdb9.png)

**Comment by nicolas33** [comment]
Does this happen with v6.5.6? v6.5.7? Configuration file? Debug logs?

**Comment by naveen1336** [comment]
this is v6.5.6rc4

**Comment by silasdb** [comment]
Hello! It also happened to me in 6.5.6. I had two computers. I could sync the messages in the former but not in the later. It didn't make any difference if I deleted that mailbox and downloaded everything again. So, looking at the first computer I found that the "problematic message" was a corrupted one, without headers and with junk data (some gzip), that was probably the result of a filesystem crash. It probably uploaded it to the mail server (a local one). So, after I deleted it in the machine where offlineimap worked, I was able to make the other machine sync IMAP.

I may provide you the "problematic message" if you want. Unfortunatelly, I didn't save the stacktrace, but I know the last messages were:

ERROR: Copying message 42422 [acc: account_name]
  list.remove(x): x not in list

Thanks!

**Comment by nicolas33** [comment]
Really nice! Now, we have solid context to find the implementation issue.

So, I believe there is a mismatch between the server and imaplib2 about the message length.

Thank you!

**Comment by nicolas33** [comment]
If someone hit this issue, please run with IMAP debug logs enabled. Thanks.

**Comment by scampi** [comment]
I am running the version 6.5.7. I got the same issue at first as OP, but then got this one, and now it seems to be going fine... (there is a lot of mail to retrieve though). Maybe this log will help you somehow.

```ini
[general]
accounts = gmail

[Account gmail]
localrepository = Local
remoterepository = Remote
status_backend = sqlite

[Repository Local]
type = Maildir
localfolders = ~/.mail/gmail

[Repository Remote]
type = Gmail
remoteuser = ***
remotepass = ***
sslcacertfile = /etc/ssl/certs/ca-certificates.crt
folderfilter = lambda folder: folder not in ['[Gmail]/All Mail', '[Gmail]/Bin', '[Gmail]/Drafts', '[Gmail]/Spam', 'SentMail', 'Trash', 'AllMail', 'Queue', 'Drafts']
```

```log
2015-10-24 20:16:50 DEBUG: [imap]:   16:50.09 Copy message from Remote:[Gmail]/Important [sync] EXAMINE ('[Gmail]/Important',)
2015-10-24 20:16:50 DEBUG: [imap]:   16:50.13 imap.gmail.com reader socket error: <class 'socket.error'> - [Errno 104] Connection reset by peer
2015-10-24 20:16:50 DEBUG: [imap]:   16:50.13 imap.gmail.com reader finished
2015-10-24 20:16:50 DEBUG: [imap]:   16:50.13 imap.gmail.com handler terminating: "socket error: <class 'socket.error'> - [Errno 104] Connection reset by peer"
2015-10-24 20:16:50 DEBUG: [imap]:   16:50.13 imap.gmail.com handler EXAMINE:MNPA92.ready.set
2015-10-24 20:16:50 DEBUG: [imap]:   16:50.13 imap.gmail.com writer finished
2015-10-24 20:16:50 DEBUG: [imap]:   16:50.13 imap.gmail.com handler state_change_free.set
2015-10-24 20:16:50 DEBUG: [imap]:   16:50.13 Copy message from Remote:[Gmail]/Important state_change_pending.release
2015-10-24 20:16:50 DEBUG: [imap]:   16:50.13 imap.gmail.com handler finished
2015-10-24 20:16:50 DEBUG: [imap]:   16:50.13 Copy message from Remote:[Gmail]/Important state => LOGOUT
2015-10-24 20:16:50 DEBUG: [imap]:   16:50.13 Copy message from Remote:[Gmail]/Important [sync] LOGOUT ()
2015-10-24 20:16:50 DEBUG: [imap]:   16:50.13 Copy message from Remote:[Gmail]/Important state_change_pending.acquire
2015-10-24 20:16:50 DEBUG: [imap]:   16:50.13 Copy message from Remote:[Gmail]/Important ["<class 'offlineimap.imaplib2.abort'>: connection closed"]
2015-10-24 20:16:50 DEBUG: [imap]:   16:50.13 Copy message from Remote:[Gmail]/Important _close_threads
2015-10-24 20:16:50 DEBUG: [imap]:   16:50.13 Copy message from Remote:[Gmail]/Important call shutdown
2015-10-24 20:16:50 DEBUG: [imap]:   16:50.13 Copy message from Remote:[Gmail]/Important state_change_pending.release
2015-10-24 20:16:50 DEBUG: [imap]:   16:50.13 Copy message from Remote:[Gmail]/Important connection closed
2015-10-24 20:16:50 DEBUG: [thread]: Register new thread 'Copy message from Remote:[Gmail]/Sent Mail' (account 'gmail')
2015-10-24 20:16:50 INFO: Copy message 26 (23 of 2426) Remote:[Gmail]/Sent Mail -> Local
2015-10-24 20:16:50 INFO: Establishing connection to imap.gmail.com:993
2015-10-24 20:16:50 DEBUG: [imap]:   16:50.14 Copy message from Remote:[Gmail]/Sent Mail imaplib2 version 2.43
2015-10-24 20:16:50 DEBUG: [imap]:   16:50.14 Copy message from Remote:[Gmail]/Sent Mail imaplib2 debug level 5, buffer level 3
2015-10-24 20:16:50 WARNING: Thread 'Copy message from Remote:[Gmail]/Important' terminated with exception:
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/offlineimap/threadutil.py", line 163, in run
    Thread.run(self)
  File "/usr/lib/python2.7/threading.py", line 763, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/usr/lib/python2.7/site-packages/offlineimap/folder/Gmail.py", line 291, in copymessageto
    super(GmailFolder, self).copymessageto(uid, dstfolder, statusfolder, register)
  File "/usr/lib/python2.7/site-packages/offlineimap/folder/Base.py", line 770, in copymessageto
    message = self.getmessage(uid)
  File "/usr/lib/python2.7/site-packages/offlineimap/folder/Gmail.py", line 77, in getmessage
    data = self._fetch_from_imap(imapobj, str(uid), 2)
  File "/usr/lib/python2.7/site-packages/offlineimap/folder/IMAP.py", line 684, in _fetch_from_imap
    imapobj.select(self.getfullname(), readonly = True)
  File "/usr/lib/python2.7/site-packages/offlineimap/imaplibutil.py", line 50, in select
    result = super(UsefulIMAPMixIn, self).select(mailbox, readonly)
  File "/usr/lib/python2.7/site-packages/offlineimap/imaplib2.py", line 1008, in select
    typ, dat = rqb.get_response('command: %s => %%s' % rqb.name)
  File "/usr/lib/python2.7/site-packages/offlineimap/imaplib2.py", line 188, in get_response
    raise typ(exc_fmt % str(val))
OfflineImapError: Server 'imap.gmail.com' closed connection, error on SELECT '[Gmail]/Important'. Server said: command: EXAMINE => socket error: <class 'socket.error'> - [Errno 104] Connection reset by peer


Last 15 debug messages logged for Copy message from Remote:[Gmail]/Important prior to exception:
thread: Register new thread 'Copy message from Remote:[Gmail]/Important' (account 'gmail')
imap:   16:50.09 Copy message from Remote:[Gmail]/Important [sync] EXAMINE ('[Gmail]/Important',)
imap:   16:50.09 Copy message from Remote:[Gmail]/Important state_change_pending.acquire
imap:   16:50.09 Copy message from Remote:[Gmail]/Important _request_push(MNPA92, EXAMINE, {}) = MNPA92
imap:   16:50.10 Copy message from Remote:[Gmail]/Important data=MNPA92 EXAMINE "[Gmail]/Important"
imap:   16:50.10 Copy message from Remote:[Gmail]/Important EXAMINE:MNPA92.ready.wait
imap:   16:50.13 Copy message from Remote:[Gmail]/Important state_change_pending.release
imap:   16:50.13 Copy message from Remote:[Gmail]/Important state => LOGOUT
imap:   16:50.13 Copy message from Remote:[Gmail]/Important [sync] LOGOUT ()
imap:   16:50.13 Copy message from Remote:[Gmail]/Important state_change_pending.acquire
imap:   16:50.13 Copy message from Remote:[Gmail]/Important ["<class 'offlineimap.imaplib2.abort'>: connection closed"]
imap:   16:50.13 Copy message from Remote:[Gmail]/Important _close_threads
imap:   16:50.13 Copy message from Remote:[Gmail]/Important call shutdown
imap:   16:50.13 Copy message from Remote:[Gmail]/Important state_change_pending.release
imap:   16:50.13 Copy message from Remote:[Gmail]/Important connection closed
```

**Comment by scampi** [comment]
Got it again. Here are the debug logs for the "thread" type.

```log
 ERROR: command: UID => socket error: <class 'ssl.SSLError'> - [SSL: DECRYPTION_FAILED_OR_BAD_RECORD_MAC] decryption failed or bad record mac (_ssl.c:1752)
 ['  File "/usr/lib/python2.7/site-packages/offlineimap/folder/IMAP.py", line 685, in _fetch_from_imap\n    res_type, data = imapobj.uid(\'fetch\', uids, query)\n', '  File "/usr/lib/python2.7/site-packages/offlineimap/imaplib2.py", line 1176, in uid\n    return self._simple_command(\'UID\', command, *args, **kw)\n', '  File "/usr/lib/python2.7/site-packages/offlineimap/imaplib2.py", line 1633, in _simple_command\n    return self._command_complete(self._command(name, *args), kw)\n', '  File "/usr/lib/python2.7/site-packages/offlineimap/imaplib2.py", line 1393, in _command_complete\n    typ, dat = rqb.get_response(\'command: %s => %%s\' % rqb.name)\n', '  File "/usr/lib/python2.7/site-packages/offlineimap/imaplib2.py", line 188, in get_response\n    raise typ(exc_fmt % str(val))\n']
 ERROR: Copying message 662 [acc: gmail]
  list.remove(x): x not in list
 ['  File "/usr/lib/python2.7/site-packages/offlineimap/folder/Base.py", line 770, in copymessageto\n    message = self.getmessage(uid)\n', '  File "/usr/lib/python2.7/site-packages/offlineimap/folder/Gmail.py", line 79, in getmessage\n    self.imapserver.releaseconnection(imapobj)\n', '  File "/usr/lib/python2.7/site-packages/offlineimap/imapserver.py", line 162, in releaseconnection\n    self.assignedconnections.remove(connection)\n']
Thread 'Copy message from Remote:[Gmail]/Sent Mail' terminated with exception:
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/offlineimap/threadutil.py", line 163, in run
    Thread.run(self)
  File "/usr/lib/python2.7/threading.py", line 763, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/usr/lib/python2.7/site-packages/offlineimap/folder/Gmail.py", line 291, in copymessageto
    super(GmailFolder, self).copymessageto(uid, dstfolder, statusfolder, register)
  File "/usr/lib/python2.7/site-packages/offlineimap/folder/Base.py", line 770, in copymessageto
    message = self.getmessage(uid)
  File "/usr/lib/python2.7/site-packages/offlineimap/folder/Gmail.py", line 79, in getmessage
    self.imapserver.releaseconnection(imapobj)
  File "/usr/lib/python2.7/site-packages/offlineimap/imapserver.py", line 162, in releaseconnection
    self.assignedconnections.remove(connection)
ValueError: list.remove(x): x not in list


Last 3 debug messages logged for Copy message from Remote:[Gmail]/Sent Mail prior to exception:
thread: Register new thread 'Copy message from Remote:[Gmail]/Sent Mail' (account 'gmail')
imap: Attempting PLAIN authentication
imap: __plainhandler: returning stephane.campinas@gmail.compwqowrjgfiooiyan
ERROR: Exceptions occurred during the run!
ERROR: command: UID => socket error: <class 'ssl.SSLError'> - [SSL: DECRYPTION_FAILED_OR_BAD_RECORD_MAC] decryption failed or bad record mac (_ssl.c:1752)

Traceback:
  File "/usr/lib/python2.7/site-packages/offlineimap/folder/IMAP.py", line 685, in _fetch_from_imap
    res_type, data = imapobj.uid('fetch', uids, query)
  File "/usr/lib/python2.7/site-packages/offlineimap/imaplib2.py", line 1176, in uid
    return self._simple_command('UID', command, *args, **kw)
  File "/usr/lib/python2.7/site-packages/offlineimap/imaplib2.py", line 1633, in _simple_command
    return self._command_complete(self._command(name, *args), kw)
  File "/usr/lib/python2.7/site-packages/offlineimap/imaplib2.py", line 1393, in _command_complete
    typ, dat = rqb.get_response('command: %s => %%s' % rqb.name)
  File "/usr/lib/python2.7/site-packages/offlineimap/imaplib2.py", line 188, in get_response
    raise typ(exc_fmt % str(val))

ERROR: Copying message 662 [acc: gmail]
  list.remove(x): x not in list

Traceback:
  File "/usr/lib/python2.7/site-packages/offlineimap/folder/Base.py", line 770, in copymessageto
    message = self.getmessage(uid)
  File "/usr/lib/python2.7/site-packages/offlineimap/folder/Gmail.py", line 79, in getmessage
    self.imapserver.releaseconnection(imapobj)
  File "/usr/lib/python2.7/site-packages/offlineimap/imapserver.py", line 162, in releaseconnection
    self.assignedconnections.remove(connection)
```

**Comment by vaab** [comment]
Hi, I have a similar issue. Although my configuration is quite different:
- I'm syncing through IMAP SSL (first runs of offline imap, so lots of downloads)
- My IMAP server is dovecot (1:2.0.19-0ubuntu2)
- Offlineimap is 6.5.7
- getting the traceback about ``list.remove(x): x not in list`` quite quickly while downloading message, this stops everything, but it seemed that re-launching offline imaps would continue. So I had to resort to a nasty ``while true; do offlineimap; sleep 1; done``, to get the job continuing.

The following logs where collected thanks to ``offlineimap -d imap``:

```
imap.kalysto.org handler:
 [imap]:   28:43.89 imap.kalysto.org handler AUTHENTICATE:GFCG2.ready.set
Account sync Kalysto:
 [imap]:   28:43.89 Account sync Kalysto state_change_pending.release
 [imap]:   28:43.89 Account sync Kalysto [async] CAPABILITY ()
 [imap]:   28:43.89 Account sync Kalysto state_change_pending.acquire
 [imap]:   28:43.89 Account sync Kalysto state_change_pending.release
 [imap]:   28:43.89 Account sync Kalysto _request_push(GFCG5, CAPABILITY, {}) = GFCG5
 [imap]:   28:43.89 Account sync Kalysto data=GFCG5 CAPABILITY
 [imap]:   28:43.89 Account sync Kalysto CAPABILITY:GFCG5.ready.wait
imap.kalysto.org writer:
 [imap]:   28:43.89 imap.kalysto.org writer > GFCG5 CAPABILITY\r\n
imap.kalysto.org reader:
 [imap]:   28:44.25 imap.kalysto.org reader poll => [(4, 1)]
 [imap]:   28:44.25 imap.kalysto.org reader rcvd 302
 [imap]:   28:44.25 imap.kalysto.org reader < * CAPABILITY IMAP4rev1 LITERAL+ SASL-IR LOGIN-REFERRALS ID ENABLE IDLE SORT SORT=DISPLAY THREAD=REFERENCES THREAD=REFS MULTIAPPEND UNSELECT CHILDREN NAMESPACE UIDPLUS LIST-EXTENDED I18NLEVEL=1 CONDSTORE QRESYNC ESEARCH ESORT SEARCHRES WITHIN CONTEXT=SEARCH LIST-STATUS\r\n
 [imap]:   28:44.25 imap.kalysto.org reader < GFCG5 OK Capability completed.\r\n
imap.kalysto.org handler:
 [imap]:   28:44.26 imap.kalysto.org handler _put_response(* CAPABILITY IMAP4rev1 LITERAL+ SASL-IR LOGIN-REFERRALS ID ENABLE IDLE SORT SORT=DISPLAY THREAD=REFERENCES THREAD=REFS MULTIAPPEND UNSELECT CHILDREN NAMESPACE UIDPLUS LIST-EXTENDED I18NLEVEL=1 CONDSTORE QRESYNC ESEARCH ESORT SEARCHRES WITHIN CONTEXT=SEARCH LIST-STATUS)
 [imap]:   28:44.26 imap.kalysto.org handler untagged_responses[CAPABILITY] 1 += ["IMAP4rev1 LITERAL+ SASL-IR LOGIN-REFERRALS ID ENABLE IDLE SORT SORT=DISPLAY THREAD=REFERENCES THREAD=REFS MULTIAPPEND UNSELECT CHILDREN NAMESPACE UIDPLUS LIST-EXTENDED I18NLEVEL=1 CONDSTORE QRESYNC ESEARCH ESORT SEARCHRES WITHIN CONTEXT=SEARCH LIST-STATUS"]
 [imap]:   28:44.26 imap.kalysto.org handler _put_response(GFCG5 OK Capability completed.)
 [imap]:   28:44.26 imap.kalysto.org handler state_change_free.set
 [imap]:   28:44.26 imap.kalysto.org handler _request_pop(GFCG5, ('OK', ['Capability completed.'])) = GFCG5
 [imap]:   28:44.26 imap.kalysto.org handler CAPABILITY:GFCG5.ready.set
Account sync Kalysto:
 [imap]:   28:44.26 Account sync Kalysto _get_untagged_response(CAPABILITY) => ['IMAP4rev1 LITERAL+ SASL-IR LOGIN-REFERRALS ID ENABLE IDLE SORT SORT=DISPLAY THREAD=REFERENCES THREAD=REFS MULTIAPPEND UNSELECT CHILDREN NAMESPACE UIDPLUS LIST-EXTENDED I18NLEVEL=1 CONDSTORE QRESYNC ESEARCH ESORT SEARCHRES WITHIN CONTEXT=SEARCH LIST-STATUS', 'IMAP4rev1 LITERAL+ SASL-IR LOGIN-REFERRALS ID ENABLE IDLE SORT SORT=DISPLAY THREAD=REFERENCES THREAD=REFS MULTIAPPEND UNSELECT CHILDREN NAMESPACE UIDPLUS LIST-EXTENDED I18NLEVEL=1 CONDSTORE QRESYNC ESEARCH ESORT SEARCHRES WITHIN CONTEXT=SEARCH LIST-STATUS']
 [imap]:   28:44.26 Account sync Kalysto _untagged_response(OK, ?, CAPABILITY) => ['IMAP4rev1 LITERAL+ SASL-IR LOGIN-REFERRALS ID ENABLE IDLE SORT SORT=DISPLAY THREAD=REFERENCES THREAD=REFS MULTIAPPEND UNSELECT CHILDREN NAMESPACE UIDPLUS LIST-EXTENDED I18NLEVEL=1 CONDSTORE QRESYNC ESEARCH ESORT SEARCHRES WITHIN CONTEXT=SEARCH LIST-STATUS', 'IMAP4rev1 LITERAL+ SASL-IR LOGIN-REFERRALS ID ENABLE IDLE SORT SORT=DISPLAY THREAD=REFERENCES THREAD=REFS MULTIAPPEND UNSELECT CHILDREN NAMESPACE UIDPLUS LIST-EXTENDED I18NLEVEL=1 CONDSTORE QRESYNC ESEARCH ESORT SEARCHRES WITHIN CONTEXT=SEARCH LIST-STATUS']
 ERROR: command: UID => connection terminated
 ['  File "/home/vaab/dev/python/offlineimap/offlineimap/folder/IMAP.py", line 685, in _fetch_from_imap\n    res_type, data = imapobj.uid(\'fetch\', uids, query)\n', '  File "/home/vaab/dev/python/offlineimap/offlineimap/imaplib2.py", line 1176, in uid\n    return self._simple_command(\'UID\', command, *args, **kw)\n', '  File "/home/vaab/dev/python/offlineimap/offlineimap/imaplib2.py", line 1633, in _simple_command\n    return self._command_complete(self._command(name, *args), kw)\n', '  File "/home/vaab/dev/python/offlineimap/offlineimap/imaplib2.py", line 1393, in _command_complete\n    typ, dat = rqb.get_response(\'command: %s => %%s\' % rqb.name)\n', '  File "/home/vaab/dev/python/offlineimap/offlineimap/imaplib2.py", line 188, in get_response\n    raise typ(exc_fmt % str(val))\n']
 ERROR: Copying message 1915 [acc: Kalysto]
  list.remove(x): x not in list
 ['  File "/home/vaab/dev/python/offlineimap/offlineimap/folder/Base.py", line 770, in copymessageto\n    message = self.getmessage(uid)\n', '  File "/home/vaab/dev/python/offlineimap/offlineimap/folder/IMAP.py", line 280, in getmessage\n    self.imapserver.releaseconnection(imapobj)\n', '  File "/home/vaab/dev/python/offlineimap/offlineimap/imapserver.py", line 162, in releaseconnection\n    self.assignedconnections.remove(connection)\n']
 ERROR: Syncing folder Informations.ml.lxc.users [acc: Kalysto]
  list.remove(x): x not in list
 ['  File "/home/vaab/dev/python/offlineimap/offlineimap/folder/Base.py", line 979, in syncmessagesto\n    action(dstfolder, statusfolder)\n', '  File "/home/vaab/dev/python/offlineimap/offlineimap/folder/Base.py", line 853, in __syncmessagesto_copy\n    register = 0)\n', '  File "/home/vaab/dev/python/offlineimap/offlineimap/folder/Base.py", line 770, in copymessageto\n    message = self.getmessage(uid)\n', '  File "/home/vaab/dev/python/offlineimap/offlineimap/folder/IMAP.py", line 280, in getmessage\n    self.imapserver.releaseconnection(imapobj)\n', '  File "/home/vaab/dev/python/offlineimap/offlineimap/imapserver.py", line 162, in releaseconnection\n    self.assignedconnections.remove(connection)\n']
 ERROR: ERROR in syncfolder for Kalysto folder Informations.ml.lxc.users: Traceback (most recent call last):
  File "/home/vaab/dev/python/offlineimap/offlineimap/accounts.py", line 567, in syncfolder
    remotefolder.syncmessagesto(localfolder, statusfolder)
  File "/home/vaab/dev/python/offlineimap/offlineimap/folder/Base.py", line 979, in syncmessagesto
    action(dstfolder, statusfolder)
  File "/home/vaab/dev/python/offlineimap/offlineimap/folder/Base.py", line 853, in __syncmessagesto_copy
    register = 0)
  File "/home/vaab/dev/python/offlineimap/offlineimap/folder/Base.py", line 770, in copymessageto
    message = self.getmessage(uid)
  File "/home/vaab/dev/python/offlineimap/offlineimap/folder/IMAP.py", line 280, in getmessage
    self.imapserver.releaseconnection(imapobj)
  File "/home/vaab/dev/python/offlineimap/offlineimap/imapserver.py", line 162, in releaseconnection
    self.assignedconnections.remove(connection)
ValueError: list.remove(x): x not in list

  list.remove(x): x not in list
```

**Comment by nicolas33** [comment]
So, it looks like the real cause is that openssl won't decrypt the data for some reason. Try with different `ssl_version` options in your configuration file. I would set it to `ssl23`.

What are your versions of OpenSSL?

**Comment by nicolas33** [comment]
Might be related to this: http://stackoverflow.com/questions/3724900/python-ssl-problem-with-multiprocessing since threading does concurrency... I can't remember if we open a new connection for each thread.

**Comment by vaab** [comment]
openssl version: 1.0.1f-1ubuntu11.4,
I changed ``ssl_version`` to ``ssl23``
I'm now on the last ``master`` (git describe gives 'v6.6.0-rc2-2-gecd9e66', I have my password patch from #266 only ) 
Please note that the hacky ``while true; do offlineimap; done`` finally reached completion, which means no specific mail was triggering the exception.
As a consequence it becomes harder for me to reproduce the bug, but nevertheless it was reproduced only by fetching new mails.
I must add also that on the server side I have some random issues that might trigger errors from dovecot... but this shouldn't trigger this buggy behavior on ``offlineimap`` side anyway.

```
imap.kalysto.org writer:
 [imap]:   10:01.90 imap.kalysto.org writer > JHHN5 CAPABILITY\r\n
imap.kalysto.org reader:
 [imap]:   10:02.26 imap.kalysto.org reader poll => [(4, 1)]
 [imap]:   10:02.26 imap.kalysto.org reader rcvd 302
 [imap]:   10:02.26 imap.kalysto.org reader < * CAPABILITY IMAP4rev1 LITERAL+ SASL-IR LOGIN-REFERRALS ID ENABLE IDLE SORT SORT=DISPLAY THREAD=REFERENCES THREAD=REFS MULTIAPPEND UNSELECT CHILDREN NAMESPACE UIDPLUS LIST-EXTENDED I18NLEVEL=1 CONDSTORE QRESYNC ESEARCH ESORT SEARCHRES WITHIN CONTEXT=SEARCH LIST-STATUS\r\n
 [imap]:   10:02.26 imap.kalysto.org reader < JHHN5 OK Capability completed.\r\n
imap.kalysto.org handler:
 [imap]:   10:02.27 imap.kalysto.org handler _put_response(* CAPABILITY IMAP4rev1 LITERAL+ SASL-IR LOGIN-REFERRALS ID ENABLE IDLE SORT SORT=DISPLAY THREAD=REFERENCES THREAD=REFS MULTIAPPEND UNSELECT CHILDREN NAMESPACE UIDPLUS LIST-EXTENDED I18NLEVEL=1 CONDSTORE QRESYNC ESEARCH ESORT SEARCHRES WITHIN CONTEXT=SEARCH LIST-STATUS)
 [imap]:   10:02.27 imap.kalysto.org handler untagged_responses[CAPABILITY] 1 += ["IMAP4rev1 LITERAL+ SASL-IR LOGIN-REFERRALS ID ENABLE IDLE SORT SORT=DISPLAY THRE"]
 [imap]:   10:02.27 imap.kalysto.org handler _put_response(JHHN5 OK Capability completed.)
 [imap]:   10:02.27 imap.kalysto.org handler _request_pop(JHHN5, ('OK', ['Capability completed.'])) [0] = JHHN5
 [imap]:   10:02.27 imap.kalysto.org handler CAPABILITY:JHHN5.ready.set
 [imap]:   10:02.27 imap.kalysto.org handler state_change_free.set
Account sync Kalysto:
 [imap]:   10:02.27 Account sync Kalysto _get_untagged_response(CAPABILITY) => ['IMAP4rev1 LITERAL+ SASL-IR LOGIN-REFERRALS ID ENABLE IDLE SORT SORT=DISPLAY TH
 [imap]:   10:02.27 Account sync Kalysto _untagged_response(OK, ?, CAPABILITY) => ['IMAP4rev1 LITERAL+ SASL-IR LOGIN-REFERRALS ID ENABLE IDLE SORT SORT=DISPLAY TH
 ERROR: command: UID => connection terminated
 ['  File "/home/vaab/dev/python/offlineimap/offlineimap/folder/IMAP.py", line 692, in _fetch_from_imap\n    res_type, data = imapobj.uid(\'fetch\', uids, query)\n', '  File "/home/vaab/dev/python/offlineimap/offlineimap/imaplib2.py", line 1220, in uid\n    return self._simple_command(\'UID\', command, *args, **kw)\n', '  File "/home/vaab/dev/python/offlineimap/offlineimap/imaplib2.py", line 1690, in _simple_command\n    return self._command_complete(self._command(name, *args), kw)\n', '  File "/home/vaab/dev/python/offlineimap/offlineimap/imaplib2.py", line 1440, in _command_complete\n    typ, dat = rqb.get_response(\'command: %s => %%s\' % rqb.name)\n', '  File "/home/vaab/dev/python/offlineimap/offlineimap/imaplib2.py", line 198, in get_response\n    raise typ(exc_fmt % str(val))\n']
 ERROR: Copying message 1915 [acc: Kalysto]
  list.remove(x): x not in list
 ['  File "/home/vaab/dev/python/offlineimap/offlineimap/folder/Base.py", line 775, in copymessageto\n    message = self.getmessage(uid)\n', '  File "/home/vaab/dev/python/offlineimap/offlineimap/folder/IMAP.py", line 287, in getmessage\n    self.imapserver.releaseconnection(imapobj)\n', '  File "/home/vaab/dev/python/offlineimap/offlineimap/imapserver.py", line 173, in releaseconnection\n    self.assignedconnections.remove(connection)\n']
 ERROR: Syncing folder Informations.ml.lxc.users [acc: Kalysto]
  list.remove(x): x not in list
 ['  File "/home/vaab/dev/python/offlineimap/offlineimap/folder/Base.py", line 995, in syncmessagesto\n    action(dstfolder, statusfolder)\n', '  File "/home/vaab/dev/python/offlineimap/offlineimap/folder/Base.py", line 864, in __syncmessagesto_copy\n    register = 0)\n', '  File "/home/vaab/dev/python/offlineimap/offlineimap/folder/Base.py", line 775, in copymessageto\n    message = self.getmessage(uid)\n', '  File "/home/vaab/dev/python/offlineimap/offlineimap/folder/IMAP.py", line 287, in getmessage\n    self.imapserver.releaseconnection(imapobj)\n', '  File "/home/vaab/dev/python/offlineimap/offlineimap/imapserver.py", line 173, in releaseconnection\n    self.assignedconnections.remove(connection)\n']
 ERROR: ERROR in syncfolder for Kalysto folder Informations.ml.lxc.users: Traceback (most recent call last):
  File "/home/vaab/dev/python/offlineimap/offlineimap/accounts.py", line 567, in syncfolder
    remotefolder.syncmessagesto(localfolder, statusfolder)
  File "/home/vaab/dev/python/offlineimap/offlineimap/folder/Base.py", line 995, in syncmessagesto
    action(dstfolder, statusfolder)
  File "/home/vaab/dev/python/offlineimap/offlineimap/folder/Base.py", line 864, in __syncmessagesto_copy
    register = 0)
  File "/home/vaab/dev/python/offlineimap/offlineimap/folder/Base.py", line 775, in copymessageto
    message = self.getmessage(uid)
  File "/home/vaab/dev/python/offlineimap/offlineimap/folder/IMAP.py", line 287, in getmessage
    self.imapserver.releaseconnection(imapobj)
  File "/home/vaab/dev/python/offlineimap/offlineimap/imapserver.py", line 173, in releaseconnection
    self.assignedconnections.remove(connection)
ValueError: list.remove(x): x not in list

  list.remove(x): x not in list
```

**Comment by nicolas33** [comment]
AFAIU, above logs are with v6.6.0-rc2-2-gecd9e66 and ssl23?

You could investigate by setting up a test environment:
* copy your configuration file;
* change the `metadata` dirrectory;
* change the `localfolders` option;
* sync.

**Comment by vaab** [comment]
> AFAIU, above logs are with v6.6.0-rc2-2-gecd9e66 and ssl23?

Yup, well, this is what I intended, something strange ?

Here's my config file:

```
[general]
accounts = Kalysto
socktimeout = 30

[Account Kalysto]
localrepository = Local
remoterepository = Kalysto

[Repository Local]
type = Maildir
localfolders = ~/.mail

[Repository Kalysto]
type = IMAP
ssl = yes
ssl_version = ssl23
remotehost = imap.kalysto.org
remoteuser = vaab
maxconnections = 10
sslcacertfile = /etc/ssl/certs/ca-certificates.crt
```

I'm quite new to ``offlineimap``, so I might have missed something obvious.

**Comment by nicolas33** [comment]
> something strange ?

Ho, no!

**Comment by vaab** [comment]
FYI, I can reproduce the bug quite reliably by forcing an error on Dovecot server side (for instance by changing the permission of a mail in the Maildir storage of the server). This force an error on the server side to a query of offline imap, and this makes it fail with the same messages than above.
So it seems that the cleanup code after an exception is bugged and try to remove twice the same connection.

**Comment by silasdb** [comment]
Hi there. After a system crash, for some reason, I cannot download one message. I'm using offlineimap version 6.5.6 under NetBSD 7.0. Using OpenSSL 1.0.2d 9 Jul 2015. I attach a file with some debug output. I'm having this issue with a Dovecot server too. I have access to the server if you need some more debugging.  Thanks!

[debug.txt](https://github.com/OfflineIMAP/offlineimap/files/33998/debug.txt)

**Comment by nicolas33** [comment]
This looks unrelated. The debug message `list.remove(x): x not in list` is due to poor error handling. 
IOW, this message **hides** the real cause. So, it's easy to mix issues with different root causes. Please, open a new one. I'll remove current comments and file. Thanks.

**Comment by vaab** [comment]
Ok, I found what was going on... The issue is in ``getmessage()`` code. For some reason it asks for a connection, runs ``_fetch_from_imap`` and with a ``try/finally`` will ask to release the connection. But ``_fetch_from_imap`` will close and require a new connection in case it fails for whatever reason to implement a retry mecanism. So when coming back to ``getmessage()`` and it's ``finally`` clause, the connection was already closed and removed from the list of available connections.

Independently to this bug, Dovecot won't be very kind upon error on his side, and as a result at least on my setup, it'll close abruptly the connection instead of giving a nice IMAP error. As a consequence, the bogus retry loop is activated quite often. 

I've corrected the retry loop in PR #279
